In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import json

In [2]:
file_template = "./virus_proteins_pdb_template_coverage.csv"
file_structure = "./virus_proteins_pdb_structure_coverage.csv"
df_template = pd.read_csv(file_template)
df_structure = pd.read_csv(file_structure)

In [3]:
df_structure

header  protein_id   Sort pept_cat  protlen  ef_coverage  \
0       CAA52564.1_4563  CAA52564.1   4563  protein        1          0.0   
1      AIW68525.1_11911  AIW68525.1  11911  protein        7          0.0   
2      AIW68499.1_11911  AIW68499.1  11911  protein        8          0.0   
3      AOO54533.1_11976  AOO54533.1  11976  protein        8          0.0   
4       AAA67112.1_4989  AAA67112.1   4989  protein        9          0.0   
...                 ...         ...    ...      ...      ...          ...   
71269  ASO76148.1_10247  ASO76148.1  10247  protein     7861       1488.0   
71270  QRC47047.1_10244  QRC47047.1  10244  protein     8043       1357.0   
71271  AIJ50565.1_10246  AIJ50565.1  10246  protein     8108       2219.0   
71272  QRC47039.1_10254  QRC47039.1  10254  protein     8927       1392.0   
71273  AYD75846.1_10227  AYD75846.1  10227  protein    13556        732.0   

       cf_coverage  ef_pdb_structure_coverage  cf_pdb_structure_coverage  
0              0.0                        0.0                        0.0  
1              0.0                        0.0                        0.0  
2              0.0                        0.0                        0.0  
3              0.0                        0.0                        0.0  
4              0.0                        0.0                        0.0  
...            ...                        ...                        ...  
71269       1488.0                        0.0                        0.0  
71270       1357.0                        0.0                        0.0  
71271       2219.0                        0.0                        0.0  
71272       1392.0                        0.0                        0.0  
71273        732.0                        0.0                        0.0  

[71274 rows x 9 columns]

In [5]:
df_template = df_template[['header', 'ef_pdb_template_coverage', 'cf_pdb_template_coverage']]

In [6]:
# merge df_template and df_structure
df = pd.merge(df_structure, df_template, on='header')

In [7]:
df

header  protein_id   Sort pept_cat  protlen  ef_coverage  \
0       CAA52564.1_4563  CAA52564.1   4563  protein        1          0.0   
1      AIW68525.1_11911  AIW68525.1  11911  protein        7          0.0   
2      AIW68499.1_11911  AIW68499.1  11911  protein        8          0.0   
3      AOO54533.1_11976  AOO54533.1  11976  protein        8          0.0   
4       AAA67112.1_4989  AAA67112.1   4989  protein        9          0.0   
...                 ...         ...    ...      ...      ...          ...   
71269  ASO76148.1_10247  ASO76148.1  10247  protein     7861       1488.0   
71270  QRC47047.1_10244  QRC47047.1  10244  protein     8043       1357.0   
71271  AIJ50565.1_10246  AIJ50565.1  10246  protein     8108       2219.0   
71272  QRC47039.1_10254  QRC47039.1  10254  protein     8927       1392.0   
71273  AYD75846.1_10227  AYD75846.1  10227  protein    13556        732.0   

       cf_coverage  ef_pdb_structure_coverage  cf_pdb_structure_coverage  \
0              0.0                        0.0                        0.0   
1              0.0                        0.0                        0.0   
2              0.0                        0.0                        0.0   
3              0.0                        0.0                        0.0   
4              0.0                        0.0                        0.0   
...            ...                        ...                        ...   
71269       1488.0                        0.0                        0.0   
71270       1357.0                        0.0                        0.0   
71271       2219.0                        0.0                        0.0   
71272       1392.0                        0.0                        0.0   
71273        732.0                        0.0                        0.0   

       ef_pdb_template_coverage  cf_pdb_template_coverage  
0                           0.0                       0.0  
1                           0.0                       0.0  
2                           0.0                       0.0  
3                           0.0                       0.0  
4                           0.0                       0.0  
...                         ...                       ...  
71269                     444.0                     157.0  
71270                     239.0                       0.0  
71271                     397.0                       0.0  
71272                       0.0                       0.0  
71273                       0.0                       0.0  

[71274 rows x 11 columns]

In [8]:
df.to_csv("./virus_proteins_pdb_structure_template_coverage.csv", index=False)

# Summary PDB structure coverage info

- Use JSON files:
    - ef_protein_plddt_pdb_template_coverage.json
    - cf_protein_plddt_pdb_template_coverage.json
- Use CSV file:
    - virus_proteins_ef_cf_coverage.csv
- Produce CSV file with AA structure coverage:
    - virus_proteins_pdb_structure_coverage.csv

In [11]:
file_coverage = "./virus_proteins_pdb_structure_template_coverage.csv"
df_coverage = pd.read_csv(file_coverage)
#df_coverage = df_coverage[['header', 'protein_id', 'Sort', 'pept_cat', 'protlen', 'ef_coverage', 'cf_coverage']]

In [12]:
df_coverage

header  protein_id   Sort pept_cat  protlen  ef_coverage  \
0       CAA52564.1_4563  CAA52564.1   4563  protein        1          0.0   
1      AIW68525.1_11911  AIW68525.1  11911  protein        7          0.0   
2      AIW68499.1_11911  AIW68499.1  11911  protein        8          0.0   
3      AOO54533.1_11976  AOO54533.1  11976  protein        8          0.0   
4       AAA67112.1_4989  AAA67112.1   4989  protein        9          0.0   
...                 ...         ...    ...      ...      ...          ...   
71269  ASO76148.1_10247  ASO76148.1  10247  protein     7861       1488.0   
71270  QRC47047.1_10244  QRC47047.1  10244  protein     8043       1357.0   
71271  AIJ50565.1_10246  AIJ50565.1  10246  protein     8108       2219.0   
71272  QRC47039.1_10254  QRC47039.1  10254  protein     8927       1392.0   
71273  AYD75846.1_10227  AYD75846.1  10227  protein    13556        732.0   

       cf_coverage  ef_pdb_structure_coverage  cf_pdb_structure_coverage  \
0              0.0                        0.0                        0.0   
1              0.0                        0.0                        0.0   
2              0.0                        0.0                        0.0   
3              0.0                        0.0                        0.0   
4              0.0                        0.0                        0.0   
...            ...                        ...                        ...   
71269       1488.0                        0.0                        0.0   
71270       1357.0                        0.0                        0.0   
71271       2219.0                        0.0                        0.0   
71272       1392.0                        0.0                        0.0   
71273        732.0                        0.0                        0.0   

       ef_pdb_template_coverage  cf_pdb_template_coverage  
0                           0.0                       0.0  
1                           0.0                       0.0  
2                           0.0                       0.0  
3                           0.0                       0.0  
4                           0.0                       0.0  
...                         ...                       ...  
71269                     444.0                     157.0  
71270                     239.0                       0.0  
71271                     397.0                       0.0  
71272                       0.0                       0.0  
71273                       0.0                       0.0  

[71274 rows x 11 columns]

In [21]:
df_coverage.to_csv("./virus_proteins_pdb_structure_coverage_updated.csv", index=False)

In [20]:
df_coverage[df_coverage['cf_90%_pdb_structure_coverage'] > 0]

header  protein_id   Sort pept_cat  protlen  ef_coverage  \
118     QSH72962.1_9614  QSH72962.1   9614  protein       39         39.0   
317    AAL40487.1_12009  AAL40487.1  12009  protein       43         43.0   
393    AUO15056.1_12445  AUO15056.1  12445  protein       44         44.0   
1244   AUO15133.1_12445  AUO15133.1  12445  protein       54         54.0   
1454   AFB76912.1_12016  AFB76912.1  12016  protein       56         56.0   
...                 ...         ...    ...      ...      ...          ...   
71257  AGT28262.1_10186  AGT28262.1  10186  protein     7150       6567.0   
71260  AAU06353.1_10183  AAU06353.1  10183  protein     7178       6820.0   
71261  ABN10874.1_10188  ABN10874.1  10188  protein     7182       6747.0   
71262  AAT98578.1_10182  AAT98578.1  10182  protein     7182       7182.0   
71264  AIL94214.1_10185  AIL94214.1  10185  protein     7247       6425.0   

       cf_coverage  ef_pdb_structure_coverage  cf_pdb_structure_coverage  \
118           39.0                       26.0                       26.0   
317           43.0                        0.0                        0.0   
393           44.0                       44.0                       44.0   
1244          54.0                       52.0                       52.0   
1454          56.0                        0.0                        0.0   
...            ...                        ...                        ...   
71257       6567.0                        0.0                        0.0   
71260       6820.0                      846.0                      846.0   
71261       6748.0                        0.0                        0.0   
71262       7182.0                      609.0                      609.0   
71264       6425.0                      134.0                      134.0   

       ef_pdb_template_coverage  cf_pdb_template_coverage  \
118                        25.0                      25.0   
317                        37.0                      37.0   
393                        44.0                      44.0   
1244                       54.0                      54.0   
1454                        0.0                       0.0   
...                         ...                       ...   
71257                    4135.0                    3178.0   
71260                    4139.0                    3197.0   
71261                    4294.0                    3284.0   
71262                    4272.0                    3333.0   
71264                    4303.0                    3371.0   

       ef_90%_pdb_structure_coverage  cf_90%_pdb_structure_coverage  
118                             26.0                           26.0  
317                             37.0                           37.0  
393                             44.0                           44.0  
1244                            52.0                           52.0  
1454                            20.0                           20.0  
...                              ...                            ...  
71257                          684.0                          684.0  
71260                         1302.0                         1302.0  
71261                          562.0                          562.0  
71262                          818.0                          818.0  
71264                          134.0                          134.0  

[3314 rows x 13 columns]

In [17]:
file_ef_poly = "./ef_protein_plddt_pdb_template_coverage_updated.json"
file_cf_poly = "./cf_protein_plddt_pdb_template_coverage_updated.json"

with open(file_cf_poly, "r") as f:
    poly_json = json.load(f)

In [ ]:
'''
ColabFold/ ESMFold structure coverage
pdb_id_90%_2024-02-20
'''

n = 0

for index, row in df_coverage.iterrows():

    n += 1
    header = row['header']

    plddt = poly_json[header]['plddt']
    structure = poly_json[header]['pdb_id_90%_2024-02-20']

    # remove plddt elements where template is np.nan based on index
    plddt_structure = [plddt[i] for i in range(len(plddt)) if pd.notna(structure[i])]
    plddt_structure = [f for f in plddt_structure if pd.notna(f)]

    df_coverage.at[index, 'cf_90%_pdb_structure_coverage'] = len(plddt_structure)

    print(n, header, len(plddt_structure))

In [ ]:
'''
ESMFold structure coverage
pdb_id_95%_2024-02-20
'''

n = 0

for index, row in df_coverage.iterrows():

    n += 1
    header = row['header']

    plddt = poly_json[header]['plddt']
    structure = poly_json[header]['pdb_id_95%_2024-02-20']

    # remove plddt elements where template is np.nan based on index
    plddt_structure = [plddt[i] for i in range(len(plddt)) if pd.notna(structure[i])]
    plddt_structure = [f for f in plddt_structure if pd.notna(f)]

    df_coverage.at[index, 'ef_pdb_structure_coverage'] = len(plddt_structure)

    print(n, header, len(plddt_structure))

# Summary PDB template coverage info

- Use JSON files: ef_protein_plddt_pdb_template_coverage.json and cf_protein_plddt_pdb_template_coverage.json
- Use CSV file: virus_proteins_ef_cf_coverage.csv
- Produce CSV file with AA template coverage and pLDDT scores per AA: virus_proteins_pdb_template_coverage.csv

In [5]:
import statistics

def plddt_mean(plddt: list, round_to: int = 1):
    '''
    Calculate the average pLDDT score.
    '''
    return round(statistics.mean(plddt), round_to)

def plddt_coverage_aa_number(plddt: list, lower_threshold: float,
                             upper_threshold: float, include_upper_threshold: bool = False):
    '''
    Calculate the coverage of a protein sequence in number of AA
    based on a pLDDT thresholds and return the number of AA.
    '''
    if include_upper_threshold:
        return len([i for i in plddt if i >= lower_threshold and i <= upper_threshold])
    else:
        return len([i for i in plddt if i >= lower_threshold and i < upper_threshold])

def plddt_coverage_across_default_categories(plddt: list):
    '''
    Calculate the coverage of a protein sequence in number of AA
    based on default pLDDT categories: [0, 50), [50, 70), [70, 90), [90, 100] -
    and return a list with the number of AA for each category.
    '''
    coverage = []
    coverage.append(plddt_coverage_aa_number(plddt, 0, 50))
    coverage.append(plddt_coverage_aa_number(plddt, 50, 70))
    coverage.append(plddt_coverage_aa_number(plddt, 70, 90))
    coverage.append(plddt_coverage_aa_number(plddt, 90, 100, include_upper_threshold=True))
    return coverage

In [2]:
file_coverage = "./virus_proteins_ef_cf_coverage.csv"
df_coverage = pd.read_csv(file_coverage)
df_coverage = df_coverage[['header', 'protein_id', 'Sort', 'pept_cat', 'protlen', 'ef_coverage', 'cf_coverage']]

In [34]:
df_coverage

header  protein_id   Sort pept_cat  protlen  ef_coverage  \
0       CAA52564.1_4563  CAA52564.1   4563  protein        1          0.0   
1      AIW68525.1_11911  AIW68525.1  11911  protein        7          0.0   
2      AIW68499.1_11911  AIW68499.1  11911  protein        8          0.0   
3      AOO54533.1_11976  AOO54533.1  11976  protein        8          0.0   
4       AAA67112.1_4989  AAA67112.1   4989  protein        9          0.0   
...                 ...         ...    ...      ...      ...          ...   
71269  ASO76148.1_10247  ASO76148.1  10247  protein     7861       1488.0   
71270  QRC47047.1_10244  QRC47047.1  10244  protein     8043       1357.0   
71271  AIJ50565.1_10246  AIJ50565.1  10246  protein     8108       2219.0   
71272  QRC47039.1_10254  QRC47039.1  10254  protein     8927       1392.0   
71273  AYD75846.1_10227  AYD75846.1  10227  protein    13556        732.0   

       cf_coverage  ef_pdb_template_coverage  ef_no_pdb_template_coverage  \
0              0.0                       0.0                          0.0   
1              0.0                       0.0                          0.0   
2              0.0                       0.0                          0.0   
3              0.0                       0.0                          0.0   
4              0.0                       0.0                          0.0   
...            ...                       ...                          ...   
71269       1488.0                     444.0                       1044.0   
71270       1357.0                     239.0                       1118.0   
71271       2219.0                     397.0                       1822.0   
71272       1392.0                       0.0                       1392.0   
71273        732.0                       0.0                        732.0   

       ef_pdb_template_plddt_0_50  ...  cf_pdb_template_coverage  \
0                             0.0  ...                       0.0   
1                             0.0  ...                       0.0   
2                             0.0  ...                       0.0   
3                             0.0  ...                       0.0   
4                             0.0  ...                       0.0   
...                           ...  ...                       ...   
71269                        21.0  ...                     157.0   
71270                         0.0  ...                       0.0   
71271                        28.0  ...                       0.0   
71272                         0.0  ...                       0.0   
71273                         0.0  ...                       0.0   

       cf_no_pdb_template_coverage  cf_pdb_template_plddt_0_50  \
0                              0.0                         0.0   
1                              0.0                         0.0   
2                              0.0                         0.0   
3                              0.0                         0.0   
4                              0.0                         0.0   
...                            ...                         ...   
71269                       1331.0                         0.0   
71270                       1357.0                         0.0   
71271                       2219.0                         0.0   
71272                       1392.0                         0.0   
71273                        732.0                         0.0   

       cf_pdb_template_plddt_50_70  cf_pdb_template_plddt_70_90  \
0                              0.0                          0.0   
1                              0.0                          0.0   
2                              0.0                          0.0   
3                              0.0                          0.0   
4                              0.0                          0.0   
...                            ...                          ...   
71269                          6.0                         31.0   
71270                

In [35]:
df_coverage.to_csv("./virus_proteins_pdb_template_coverage.csv", index=False)

In [31]:
file_ef_poly = "./ef_protein_plddt_pdb_template_coverage.json"
file_cf_poly = "./cf_protein_plddt_pdb_template_coverage.json"

with open(file_cf_poly, "r") as f:
    poly_json = json.load(f)

In [ ]:
'''
ColabFold template coverage
pdb_id_30%_2018-04-30
'''


n = 0

for index, row in df_coverage.iterrows():

    n += 1
    header = row['header']

    plddt = poly_json[header]['plddt']
    template = poly_json[header]['pdb_id_30%_2018-04-30']

    # remove plddt elements where template is np.nan based on index
    plddt_template = [plddt[i] for i in range(len(plddt)) if pd.notna(template[i])]
    plddt_template = [f for f in plddt_template if pd.notna(f)]

    plddt_no_template = [plddt[i] for i in range(len(plddt)) if pd.isna(template[i])]
    plddt_no_template = [f for f in plddt_no_template if pd.notna(f)]
    
    df_coverage.at[index, 'cf_pdb_template_coverage'] = len(plddt_template)
    df_coverage.at[index, 'cf_no_pdb_template_coverage'] = len(plddt_no_template)

    template_0_50, template_50_70, template_70_90, template_90_100 = plddt_coverage_across_default_categories(plddt_template)
    df_coverage.at[index, 'cf_pdb_template_plddt_0_50'] = template_0_50
    df_coverage.at[index, 'cf_pdb_template_plddt_50_70'] = template_50_70
    df_coverage.at[index, 'cf_pdb_template_plddt_70_90'] = template_70_90
    df_coverage.at[index, 'cf_pdb_template_plddt_90_100'] = template_90_100

    no_template_0_50, no_template_50_70, no_template_70_90, no_template_90_100 = plddt_coverage_across_default_categories(plddt_no_template)
    df_coverage.at[index, 'cf_no_pdb_template_plddt_0_50'] = no_template_0_50
    df_coverage.at[index, 'cf_no_pdb_template_plddt_50_70'] = no_template_50_70
    df_coverage.at[index, 'cf_no_pdb_template_plddt_70_90'] = no_template_70_90
    df_coverage.at[index, 'cf_no_pdb_template_plddt_90_100'] = no_template_90_100

    print(n, header, len(plddt_template), len(plddt_no_template))

In [ ]:
'''
ESMFold template coverage
pdb_id_30%_2020-05-01
'''


n = 0

for index, row in df_coverage.iterrows():

    n += 1
    header = row['header']

    plddt = poly_json[header]['plddt']
    template = poly_json[header]['pdb_id_30%_2020-05-01']

    # remove plddt elements where template is np.nan based on index
    plddt_template = [plddt[i] for i in range(len(plddt)) if pd.notna(template[i])]
    plddt_template = [f for f in plddt_template if pd.notna(f)]

    plddt_no_template = [plddt[i] for i in range(len(plddt)) if pd.isna(template[i])]
    plddt_no_template = [f for f in plddt_no_template if pd.notna(f)]
    
    df_coverage.at[index, 'ef_pdb_template_coverage'] = len(plddt_template)
    df_coverage.at[index, 'ef_no_pdb_template_coverage'] = len(plddt_no_template)

    template_0_50, template_50_70, template_70_90, template_90_100 = plddt_coverage_across_default_categories(plddt_template)
    df_coverage.at[index, 'ef_pdb_template_plddt_0_50'] = template_0_50
    df_coverage.at[index, 'ef_pdb_template_plddt_50_70'] = template_50_70
    df_coverage.at[index, 'ef_pdb_template_plddt_70_90'] = template_70_90
    df_coverage.at[index, 'ef_pdb_template_plddt_90_100'] = template_90_100

    no_template_0_50, no_template_50_70, no_template_70_90, no_template_90_100 = plddt_coverage_across_default_categories(plddt_no_template)
    df_coverage.at[index, 'ef_no_pdb_template_plddt_0_50'] = no_template_0_50
    df_coverage.at[index, 'ef_no_pdb_template_plddt_50_70'] = no_template_50_70
    df_coverage.at[index, 'ef_no_pdb_template_plddt_70_90'] = no_template_70_90
    df_coverage.at[index, 'ef_no_pdb_template_plddt_90_100'] = no_template_90_100

    print(n, header, len(plddt_template), len(plddt_no_template))

# Calculate PDB coverage and save it in json

- Use JSON files with protein info:
    - ./cf_protein_plddt_coverage.json
    - ./ef_protein_plddt_coverage.json
- Use JSON files with training data cut-off info:
    - ./pdb_ids_released_before_2018-04-30.json
    - ./pdb_ids_released_before_2020-05-01.json
- Use TSV files with Mmseqs2 easy-search results against PDB 2024-02-20 release:
    - ../mmseqs2_analysis_pdb/all_proteins_pdb_results.tsv
- Produce JSON files with PDB structure info (95% seq identity to PDB 2024-02-20) and tempalte info (30% seq identity to PDB 2018-04-30 and 2020-05-01 for ColabFold and ESMFold respectively):
    - ./cf_protein_plddt_pdb_template_coverage.json
    - ./ef_protein_plddt_pdb_template_coverage.json

### JSON file structure:

- 71,274 items (matches the number of protein records)
- key: protein record_id (eg. 'ASO76148.1_10247')
    - value_1 key: 'record_id'
    - value_1 value: list of record_id (eg. 'ASO76148.1_10247') that cover this protein or its region structurally
    - value_2 key: 'plddt'
    - value_2 value: list of per residue pLDDT scores (eg. 76.0)
    - value_3 key: 'pdb_id_95%_2024-02-20'
    - value_3 value: list of PDB IDs (eg. mom3) from PDB 2024-02-20 that cover this protein or its region structurally with 95% sequence identity
    - value_4 key: 'pdb_id_30%_2020-05-01' (ESMFold only)
    - value_4 value: list of PDB IDs (eg. mom3) from PDB 2020-05-01 (ESMFold training data cut-off) that cover this protein or its region structurally with 30% sequence identity (template)
    - value_5 key: 'pdb_id_30%_2018-04-30' (ColabFold only)
    - value_4 value: list of PDB IDs (eg. mom3) from PDB 2018-04-30 (ColabFold training data cut-off) that cover this protein or its region structurally with 30% sequence identity (template)

In [ ]:
'''
PDB entries released before 30 April 2018 (AlphaFold2 and AlphaFold2-Multimer training data cut-off)
PDB entries released before 1 May 2020 (ESMFold training data cut-off)
'''

#file = "./pdb_ids_released_before_2018-04-30.json"
file = "./pdb_ids_released_before_2020-05-01.json"
with open(file, "r") as f:
    pdb_ids = json.load(f)

pdb_ids_list = []
for i in pdb_ids['result_set']:
    pdb_ids_list.append(i['identifier'].lower())

print(len(pdb_ids_list))

In [3]:
file_coverage = "./virus_proteins_pdb_structure_template_coverage_updated.csv"
df_coverage = pd.read_csv(file_coverage)

In [4]:
df_coverage

header  protein_id   Sort pept_cat  protlen  ef_coverage  \
0       CAA52564.1_4563  CAA52564.1   4563  protein        1          0.0   
1      AIW68525.1_11911  AIW68525.1  11911  protein        7          0.0   
2      AIW68499.1_11911  AIW68499.1  11911  protein        8          0.0   
3      AOO54533.1_11976  AOO54533.1  11976  protein        8          0.0   
4       AAA67112.1_4989  AAA67112.1   4989  protein        9          0.0   
...                 ...         ...    ...      ...      ...          ...   
71269  ASO76148.1_10247  ASO76148.1  10247  protein     7861       1488.0   
71270  QRC47047.1_10244  QRC47047.1  10244  protein     8043       1357.0   
71271  AIJ50565.1_10246  AIJ50565.1  10246  protein     8108       2219.0   
71272  QRC47039.1_10254  QRC47039.1  10254  protein     8927       1392.0   
71273  AYD75846.1_10227  AYD75846.1  10227  protein    13556        732.0   

       cf_coverage  ef_pdb_structure_coverage  cf_pdb_structure_coverage  \
0              0.0                        0.0                        0.0   
1              0.0                        0.0                        0.0   
2              0.0                        0.0                        0.0   
3              0.0                        0.0                        0.0   
4              0.0                        0.0                        0.0   
...            ...                        ...                        ...   
71269       1488.0                        0.0                        0.0   
71270       1357.0                        0.0                        0.0   
71271       2219.0                        0.0                        0.0   
71272       1392.0                        0.0                        0.0   
71273        732.0                        0.0                        0.0   

       ef_pdb_template_coverage  cf_pdb_template_coverage  \
0                           0.0                       0.0   
1                           0.0                       0.0   
2                           0.0                       0.0   
3                           0.0                       0.0   
4                           0.0                       0.0   
...                         ...                       ...   
71269                     444.0                     157.0   
71270                     239.0                       0.0   
71271                     397.0                       0.0   
71272                       0.0                       0.0   
71273                       0.0                       0.0   

       ef_90%_pdb_structure_coverage  
0                                0.0  
1                                0.0  
2                                0.0  
3                                0.0  
4                                0.0  
...                              ...  
71269                            0.0  
71270                            0.0  
71271                            0.0  
71272                            0.0  
71273                            0.0  

[71274 rows x 12 columns]

In [5]:
# PDB 2024-02-20 release
file = "../mmseqs2_analysis_pdb/all_proteins_pdb_results.tsv" # Mmseqs2 results (size is over 400 MB)
df_pdb = pd.read_csv(file, sep="\t")
df_pdb['target_short'] = df_pdb['target'].str.split(".").str[0]

In [7]:
# Filter out the pdb_ids that are not in the list
df_pdb = df_pdb[df_pdb['target_short'].isin(pdb_ids_list)]

In [6]:
df_pdb = df_pdb[df_pdb['fident'] >= 0.9]
df_pdb.sort_values(by=['query', 'alnlen'], ascending=False, inplace=True)

In [7]:
df_pdb.head(5)

query         target  \
137527   SYZ47171.1_8840  6m7d.cif.gz_A   
1173502  QZA87341.1_7245  5mz4.cif.gz_A   
1173503  QZA87341.1_7245  5wx1.cif.gz_A   
1173519  QZA87341.1_7245  4cbl.cif.gz_B   
1173524  QZA87341.1_7245  4cbm.cif.gz_A   

                                                   theader  fident  alnlen  \
137527   6m7d.cif.gz_A Structure of ncleoprotein of sen...   0.924     412   
1173502  5mz4.cif.gz_A Crystal Structure of full-lengh ...   0.930     691   
1173503  5wx1.cif.gz_A The closed-conformation crystal ...   0.934     683   
1173519              4cbl.cif.gz_B Pestivirus NS3 helicase   0.936     489   
1173524              4cbm.cif.gz_A Pestivirus NS3 helicase   0.915     487   

         mismatch  gapopen  qstart  qend  qlen   qcov  tstart  tend  tlen  \
137527         31        0       3   414   524  0.786       1   412   412   
1173502        47        0    1586  2276  3894  0.177      34   704   704   
1173503        43        0    1586  2268  3894  0.175      34   693   693   
1173519        31        0    1787  2275  3894  0.126       1   479   479   
1173524        40        0    1789  2275  3894  0.125       2   467   467   

          tcov         evalue  bits                            taxname  \
137527   1.000  1.566000e-248   769        Sendai virus (strain Ohita)   
1173502  0.953   0.000000e+00  1308  Hog cholera virus (strain Alfort)   
1173503  0.952   0.000000e+00  1294        Classical swine fever virus   
1173519  1.000  1.415000e-265   914        Classical swine fever virus   
1173524  0.998  3.682000e-256   883        Classical swine fever virus   

        target_short  
137527          6m7d  
1173502         5mz4  
1173503         5wx1  
1173519         4cbl  
1173524         4cbm

In [8]:
file_ef_poly = "./ef_protein_plddt_pdb_template_coverage.json"
file_cf_poly = "./cf_protein_plddt_pdb_template_coverage.json"

with open(file_cf_poly, "r") as f:
    poly_json = json.load(f)

In [ ]:
'''
ESMFold coverage by PDB entries with 30% sequence identity
from the PDB release 2020-05-01
'''

n = 0

for key, value in poly_json.items():

    n += 1
    header = key
    protlen = df_coverage.loc[df_coverage['header'] == header, 'protlen'].values[0]
    pdb_id_list = [np.nan] * protlen

    if header in df_pdb['query'].values:

        df_pdb_sub = df_pdb[df_pdb['query'] == header].sort_values(by=['alnlen'], ascending=False)

        for index, row in df_pdb_sub.iterrows():

            pdb_id = row['target_short']
            start = row['qstart']
            end = row['qend']
            
            for i in range(start, end+1):
                if pdb_id_list[i-1] is np.nan:
                    try:
                        pdb_id_list[i-1] = pdb_id
                    except IndexError:
                        print('ERROR: ', header, end - start + 1, len(pdb_id_list))
                    
    value['pdb_id_30%_2020-05-01'] = pdb_id_list
    pdb_coverage = len(pdb_id_list) - pdb_id_list.count(np.nan)
    print(n, header, round(pdb_coverage/ protlen, 2), pdb_coverage)

In [17]:
# save json file
file_json_poly = "./ef_protein_plddt_pdb_template_coverage.json"
with open(file_json_poly, "w") as f:
    json.dump(poly_json, f)

In [ ]:
'''
ColabFold coverage by PDB entries with 30% sequence identity
from the PDB release 2018-04-30
'''

n = 0

for key, value in poly_json.items():

    n += 1
    header = key
    protlen = df_coverage.loc[df_coverage['header'] == header, 'protlen'].values[0]
    pdb_id_list = [np.nan] * protlen

    if header in df_pdb['query'].values:

        df_pdb_sub = df_pdb[df_pdb['query'] == header].sort_values(by=['alnlen'], ascending=False)

        for index, row in df_pdb_sub.iterrows():

            pdb_id = row['target_short']
            start = row['qstart']
            end = row['qend']
            
            for i in range(start, end+1):
                if pdb_id_list[i-1] is np.nan:
                    try:
                        pdb_id_list[i-1] = pdb_id
                    except IndexError:
                        print('ERROR: ', header, end - start + 1, len(pdb_id_list))
                     
    value['pdb_id_30%_2018-04-30'] = pdb_id_list
    pdb_coverage = len(pdb_id_list) - pdb_id_list.count(np.nan)
    print(n, header, round(pdb_coverage/ protlen, 2), pdb_coverage)

In [21]:
# save json file
file_json_poly = "./cf_protein_plddt_pdb_template_coverage.json"
with open(file_json_poly, "w") as f:
    json.dump(poly_json, f)

In [ ]:
'''
ColabFold/ ESMFold coverage by PDB entries with 90% sequence identity
from the PDB release 2024-02-20
'''

n = 0

for key, value in poly_json.items():

    n += 1
    header = key
    protlen = df_coverage.loc[df_coverage['header'] == header, 'protlen'].values[0]
    pdb_id_list = [np.nan] * protlen

    if header in df_pdb['query'].values:

        df_pdb_sub = df_pdb[df_pdb['query'] == header].sort_values(by=['alnlen'], ascending=False)

        for index, row in df_pdb_sub.iterrows():

            pdb_id = row['target_short']
            start = row['qstart']
            end = row['qend']
            
            for i in range(start, end+1):
                if pdb_id_list[i-1] is np.nan:
                    try:
                        pdb_id_list[i-1] = pdb_id
                    except IndexError:
                        print('ERROR: ', header, end - start + 1, len(pdb_id_list))
                    
        
    value['pdb_id_90%_2024-02-20'] = pdb_id_list
    pdb_coverage = len(pdb_id_list) - pdb_id_list.count(np.nan)
    df_coverage.loc[df_coverage['header'] == header, 'cf_90%_pdb_structure_coverage'] = pdb_coverage
    print(n, header, round(pdb_coverage/ protlen, 2), pdb_coverage)


In [ ]:
'''
ColabFold/ ESMFold coverage by PDB entries with 95% sequence identity
from the PDB release 2024-02-20
'''

n = 0

for key, value in poly_json.items():

    n += 1
    header = key
    protlen = df_coverage.loc[df_coverage['header'] == header, 'protlen'].values[0]
    pdb_id_list = [np.nan] * protlen

    if header in df_pdb['query'].values:

        df_pdb_sub = df_pdb[df_pdb['query'] == header].sort_values(by=['alnlen'], ascending=False)

        for index, row in df_pdb_sub.iterrows():

            pdb_id = row['target_short']
            start = row['qstart']
            end = row['qend']
            
            for i in range(start, end+1):
                if pdb_id_list[i-1] is np.nan:
                    try:
                        pdb_id_list[i-1] = pdb_id
                    except IndexError:
                        print('ERROR: ', header, end - start + 1, len(pdb_id_list))
                    
        
    value['pdb_id_95%_2024-02-20'] = pdb_id_list
    pdb_coverage = len(pdb_id_list) - pdb_id_list.count(np.nan)
    #df_coverage.loc[df_coverage['header'] == header, 'pdb_id_95%_2024-02-20'] = pdb_coverage
    print(n, header, round(pdb_coverage/ protlen, 2), pdb_coverage)


# Make one JSON file with pLDDT scores for all proteins (not records)

It had to be done only once. No need to repeat this step.

- Use JSON files for polyproteins:
    - ../plddt_scores/cf_plddt_coverage_polyproteins_from_peptides_and_regions.json
    - ../plddt_scores/ef_plddt_coverage_polyproteins_from_peptides_and_regions.json
- Use JSON files for all records:
    - ../plddt_scores/all_ESMFold_plddt_scores.json
    - ../plddt_scores/all_rank_001_CF_plddt_scores.json
- Produce JSON files for all proteins:
    - ./cf_protein_plddt_coverage.json
    - ./ef_protein_plddt_coverage.json

In [34]:
file_ef_all = "../plddt_scores/all_ESMFold_plddt_scores.json"
file_cf_all = "../plddt_scores/all_rank_001_CF_plddt_scores.json"

with open(file_cf_all, "r") as f:
    ef_all = json.load(f)

In [35]:
file_ef_poly = "../plddt_scores/ef_plddt_coverage_polyproteins_from_peptides_and_regions.json"
file_cf_poly = "../plddt_scores/cf_plddt_coverage_polyproteins_from_peptides_and_regions.json"

with open(file_cf_poly, "r") as f:
    ef_poly = json.load(f)

In [ ]:
'''
Currently poly json contains only record_id and plddt values for protein records
that were predicted as peptides or regions. This code will add the record_id and plddt
value for protein records that were fully predicted as well.
'''

n = 0

for index, row in df_coverage.iterrows():

    n += 1
    header = row['header']
    protlen = row['protlen']

    if header in ef_poly.keys():
        continue
    
    ef_poly[header] = {}

    if header not in ef_all.keys():
        record_id_list = [np.nan] * protlen
        plddt_list = [np.nan] * protlen
    else:
        record_id_list = [header] * protein_seq
        plddt_list = ef_all[header]
    
    ef_poly[header]['record_id'] = record_id_list
    ef_poly[header]['plddt'] = plddt_list

    print(n, header, protlen, record_id_list, plddt_list)

In [41]:
# save json file
file_updated = "./cf_protein_plddt_coverage.json"
with open(file_updated, "w") as f:
    json.dump(ef_poly, f)